src/stacked_borrows.rs

```diff
@@ -194,140 +197,136 @@ impl Default for Tag {
-        // from the reborrow of the first statement, and subsequently also pop the `SharedReadWrite` for `raw`.
-        // This pattern occurs a lot in the standard library: create a raw pointer, then also create a shared
-        // reference and use that.
-        {
-            // Implemented with indices because there does not seem to be a nice iterator and range-based
-            // API for this.
-            let mut cur = granting_idx + 1;
-            while let Some(item) = self.borrows.get(cur) {
-                if granting_perm.compatible_with(access, item.perm) {
-                    // Keep this, check next.
-                    cur += 1;
-                } else {
-                    // Aha! This is a bad one, remove it, and make sure it is not protected.
-                    let item = self.borrows.remove(cur);
-                    if let Some(call) = item.protector {
-                        if global.is_active(call) {
-                            return err!(MachineError(format!(
-                                "not granting {} access to tag {} because incompatible item {} is protected",
-                                access, tag, item
-                            )));
-                        }
-                    }
-                    trace!("access: removing item {}", item);
+        // Step 2: Remove incompatible items above them.  Make sure we do not remove protected
+        // items.  Behavior differs for reads and writes.
+        if access == AccessKind::Write {
+            // Remove everything above the write-compatible items, like a proper stack. This makes sure read-only and unique
+            // pointers become invalid on write accesses (ensures F2a, and ensures U2 for write accesses).
+            let first_incompatible_idx = self.find_first_write_incompaible(granting_idx);
+            while self.borrows.len() > first_incompatible_idx {
+                let item = self.borrows.pop().unwrap();
+                trace!("access: popping item {}", item);
+                Stack::check_protector(&item, Some(tag), global)?;
+            }
+        } else {
+            // On a read, *disable* all `Unique` above the granting item.  This ensures U2 for read accesses.
+            // The reason this is not following the stack discipline (by removing the first Unique and
+            // everything on top of it) is that in `let raw = &mut *x as *mut _; let _val = *x;`, the second statement
+            // would pop the `Unique` from the reborrow of the first statement, and subsequently also pop the
+            // `SharedReadWrite` for `raw`.
+            // This pattern occurs a lot in the standard library: create a raw pointer, then also create a shared
+            // reference and use that.
+            // We *disable* instead of removing `Unique` to avoid "connecting" two neighbouring blocks of SRWs.
+            for idx in (granting_idx+1 .. self.borrows.len()).rev() {
```

Comment:
```suggestion
            for item in self.borrows.iter_mut().skip(granting_idx + 1).rev() {
```

Reply:
I'd prefer to keep this code similar to the one in the other branch of the conditional.